src/tiled/stampbrush.cpp

```diff
@@ -337,6 +337,12 @@ void StampBrush::doPaint(int flags, QHash<TileLayer*, QRegion> *paintedRegions)
     if (!preview)
         return;
 
+    if(!mRandomCacheValid) {
+        updateRandomList();
+        mRandomCacheValid = true;
+        updatePreview();
```

Comment:
I think it is wrong for `doPaint` to call `updatePreview`. Maybe it would be better to check `mRandomCacheValid` in `updatePreview` instead?

Reply:
Ok. I put it in `drawPreviewLayer` so it's only checked when the stamp is being used randomly